python/tempo/interpol.py

```diff
@@ -0,0 +1,321 @@
+                )
+        for column in target_cols:
+            if column not in str(df.columns):
+                raise ValueError(
+                    f"Target Column: '{column}' does not exist in DataFrame."
+                )
+            if df.select(column).dtypes[0][1] not in supported_target_col_types:
+                raise ValueError(
+                    f"Target Column needs to be one of the following types: {supported_target_col_types}"
+                )
+
+        if ts_col not in str(df.columns):
+            raise ValueError(
+                f"Timestamp Column: '{ts_col}' does not exist in DataFrame."
+            )
+
+        if df.select(ts_col).dtypes[0][1] != "timestamp":
+            raise ValueError(f"Timestamp Column needs to be of timestamp type.")
+
+    def __fill_series(
+        self, df: DataFrame, partition_cols: List[str], ts_col: str, target_col: str
+    ) -> DataFrame:
+        """
+        Apply forward and backward fill (value and timestamp) to the timeseries dataset.
+
+        :param df  - input dataframe
+        :param partition_cols   - partition columns for window (columns must exist in the input dataframe)
+        :ts_col int - timeseries column (must exist in the input dataframe)
+        :target_col int - column to perform interpolation (must exist in the input dataframe)
+        """
+        window_ff = (
+            Window.partitionBy(*partition_cols)
+            .orderBy(ts_col)
+            .rowsBetween(-sys.maxsize, 0)
+        )
+        window_bf = (
+            Window.partitionBy(*partition_cols)
+            .orderBy(ts_col)
+            .rowsBetween(0, sys.maxsize)
+        )
+
+        read_last = last(df[target_col], ignorenulls=True).over(window_ff)
+        read_next = first(df[target_col], ignorenulls=True).over(window_bf)
+        readtime_last = last(df["surrogate_ts"], ignorenulls=True).over(window_ff)
+        readtime_next = first(df["surrogate_ts"], ignorenulls=True).over(window_bf)
+
+        filled_series: DataFrame = (
+            df.withColumn("readvalue_ff", read_last)
+            .withColumn("readvalue_bf", read_next)
+            .withColumn("readtime_ff", readtime_last)
+            .withColumn("readtime_bf", readtime_next)
+            .filter(col("readtime_bf").isNotNull())
+        )
+
+        return filled_series
+
+    def __interpolate_column(
+        self,
+        epoch_series: DataFrame,
+        partition_cols: List[str],
+        ts_col: str,
+        target_col: str,
+        fill: str,
+    ) -> DataFrame:
+        """
+        Apply interpolation to column.
+
+        :param epoch_series  - input DataFrame in epoch seconds
+        :param ts_col   - timestamp column name
+        :param target_col   - numeric column to interpolate
+        :param fill   - interpolation function to fill missing values
+        """
+        # Generate Fill for Dataset
+        filled_series: DataFrame = self.__fill_series(
```

Comment:
If we can return a list of new columns [last(col(f)) for f in target_list] then expand in the final data frame, we can avoid the joins. I'll work on a prototype for that - this would be really helpful in avoiding the join even though logically we are still going through each column to compute the last non-null value.